python/pyspark/sql/types.py

```diff
@@ -57,7 +57,25 @@ def __ne__(self, other):
 
     @classmethod
     def typeName(cls):
-        return cls.__name__[:-4].lower()
+        typeTypeNameMap = {"DataType": "data",
+                           "NullType": "null",
+                           "StringType": "string",
+                           "BinaryType": "binary",
+                           "BooleanType": "boolean",
+                           "DateType": "date",
+                           "TimestampType": "timestamp",
+                           "DecimalType": "decimal",
+                           "DoubleType": "double",
+                           "FloatType": "float",
+                           "ByteType": "byte",
+                           "IntegerType": "integer",
+                           "LongType": "long",
+                           "ShortType": "short",
+                           "ArrayType": "array",
+                           "MapType": "map",
+                           "StructField": "struct",
```

Comment:
Yeah, I don't think it is valid to call `typeName` against a `StructField`. Actually, `StructField` is not a data type, strictly speaking...

I don't know why `StructField` inherits `DataType` in pyspark. In scala, it is not.

Throwing an exception when calling `typeName` on `StructField` seems good enough, instead of a map like this.